test/model/adapters/sql_joins_test.rb

```diff
@@ -0,0 +1,311 @@
+
+      attributes :user_id, :total
+    end
+
+    class TestOrderRepository
+      include Lotus::Repository
+    end
+
+    class TestAge
+      include Lotus::Entity
+
+      attributes :value, :label
+    end
+
+    class TestAgeRepository
+      include Lotus::Repository
+    end
+
+    class TestCountry
+      include Lotus::Entity
+
+      attributes :code, :country_id
+    end
+
+    class TestCountryRepository
+      include Lotus::Repository
+    end
+
+    @mapper = Lotus::Model::Mapper.new do
+      collection :users do
+        entity TestUser
+
+        attribute :id,   Integer
+        attribute :country_id, Integer
+        attribute :name, String
+        attribute :age,  Integer
+      end
+
+      collection :devices do
+        entity TestDevice
+
+        attribute :id, Integer
+        attribute :u_id, Integer
+      end
+
+      collection :orders do
+        entity TestOrder
+
+        attribute :id,      Integer
+        attribute :user_id, Integer
+        attribute :total,   Integer
+      end
+
+      collection :ages do
+        entity TestAge
+
+        attribute :id,    Integer
+        attribute :value, Integer
+        attribute :label, String
+      end
+
+      collection :countries do
+        entity TestCountry
+
+        identity :country_id
+        attribute :id,   Integer, as: :country_id
+        attribute :code, String
+      end
+    end.load!
+
+    @adapter = Lotus::Model::Adapters::SqlAdapter.new(@mapper, SQLITE_CONNECTION_STRING)
+  end
+
+  describe '#query' do
+    let(:user1) { TestUser.new(name: 'L',  age: '32') }
+    let(:user2) { TestUser.new(name: 'MG', age: 31)   }
+    let(:user3) { TestUser.new(name: 'S',  age: 2)    }
+
+    describe 'join' do
+      describe 'inner' do
+        describe 'with an empty collection' do
+          before do
+            @adapter.clear(:orders)
+            @adapter.clear(:users)
+          end
+
+          it 'returns an empty result set' do
+            result = @adapter.query(:orders) do
+              join(:users)
+            end.all
+
+            result.must_be_empty
+          end
+        end
+
+        describe 'with a filled collection' do
+          describe 'and default options' do
+            before do
+              @adapter.clear(:users)
+              @adapter.clear(:orders)
+
+              @created_user = @adapter.create(:users, user1)
+
+              @order1 = TestOrder.new(user_id: @created_user.id, total: 100)
+              @order2 = TestOrder.new(user_id: @created_user.id, total: 200)
+              @order3 = TestOrder.new(user_id: nil,              total: 300)
+
+              @adapter.create(:orders, @order1)
+              @adapter.create(:orders, @order2)
+              @adapter.create(:orders, @order3)
+
+              TestUserRepository.adapter = @adapter
+              TestOrderRepository.adapter = @adapter
+            end
+
+            it 'returns records' do
+              created_user = TestUserRepository.first
```

Comment:
To be sure we're referring to the same record, we should use `@created_user` from the `before` block, instead of finding it again. :wink: